File: session4/LinkedList.java
//most lists are empty -- keep a low overhead
public class LinkedList {

    //a static inner class removes the reference to the parent
    private static class Node {
        public Node next;
        public int val;
    }

    private Node head;
    public LinkedList() {
        head = null;
    }

    //O(1), this is a smart way to do it, no special case
    public void addStart(int val) {
        Node temp = new Node();
        temp.val = v;
        temp.next = head;
        head = temp;
    }

    //O(n) in total
    public void addEnd(int val) {
        Node temp = new Node();
        temp.val = v;
        temp.next = null;
        if(head == null) {
            head = temp;
            return;
        }
        //guaranteed to have at least 1 element
        Node p;
        for(p = head; p.next != null; p = p.next); //move it to the end
        p.next = temp;
    }

    //O(n), but you could also put it on the class to distribute the work over
    //all the add and reduce call.  Then this would be O(1)
    public int size() {
        int count = 0;
        for(Node p = head; p != null; p = p.next)
            count++;
        return count;
    }

    //there are certain operations that cannot be corrected and will only slow us down,
    //such as preventing the access of an item not in the list
    public int get(int pos) { //O(pos) = O(n)
        Node p = head;
        for(int i = 0; i < pos; i++) {
            p = p.next;
        }
        return p.val;
    }

    //O(1)
    //once you have this API where you are returning an integer, there is no sensible way to
    //handle the error of having an empty list, let them crash
    public int removeStart() {
        Node temp = head;
        head = head.next;
        return temp.val;
    }

    //using .next.next is slow because each time we jump, we have to access two ahead, even on ones we've checked
    //again, if you have no elements, you have no business coming here
    public int removeEnd() {
        Node p = head, q = p.next;
        if(q == null) { //only have one element in the list
            int last = head.val;
            head = null;
            return last;
        }
        //invariant: at least 2 elements
        for(; q.next != null; p = q, q = q.next) {
            //q is on the last one, p is on the second to last
        }
        p.next = null;
        return q.val; //keep in mind we have a memory leak if we are not in a garbage collected language
    }

    //StringBuffer is a thread safe version of StringBuilder, but we are the only ones with our builder, so only one thread has it --
    //thus we don't care about being thread safe
    public String toString() {
        StringBuilder b = new StringBuilder(size * 6);
        for(Node p = head; p != null; p = p.next)
            b.append(p.val).append(' ');
        return b.toString();
    }
}

File: session4/LinkedList.cpp
template<typename T>
class LinkedList {
private:
    class Node {
    public:
        Node* next; //in C++, just saying "Node" would be a recursive object reference
        T val; //using a generic or template
        Node(Node* next, const T& val) : next(next) , val(v) {}
    };
    Node* head;

public:
    LinkedList() : head(nullptr) {}
    void addStart(const T& val) { //we are unsure of the size of this, don't copy it
        Node* temp = new Node(head, val);
        //replaced with the constructor
        // temp->val = val; //arrow is a dereference followed by an access
        // temp->next = head; 
        head = temp;
        //or just in one line:
        //head = new Node(head, val);
    }

    void addEnd(const T& val) {
        Node* temp = new Node(nullptr, val);
        Node* p = head;
        if(p == null) {
            head = temp;
            return;
        }
        for(; p->next != nullptr; p = p->next);
        p->next = temp;
    }

    void removeStart() {
        Node* temp = head;
        head = head->next;
        delete temp; //MUST free the memory or you will leak
    }

    void removeEnd() {
        Node* p = head;
        Node* q = p.next;
        if(q == nullptr) {
            delete head;
            head = nullptr;
            return;
        }
        for(; q->next != nullptr; p = q, q = q->next);
        delete p->next;
        p->next = nullptr;
    }

    //make a copy of a linked list
    LinkedList(const LinkedList& orig) {
        if(orig.head == nullptr) {
            head = nullptr;
            return;
        }
        head = new Node(nullptr, orig.head->val);
        Node* p = head;

        Node* src = orig.head->next;
        Node* dest = head;
        while(src != null) {
            dest->next = new Node(nullptr, src->val);
            dest = dest->next;
            src = src->next;
        }
    }

    //the destructor
    ~LinkedList() {
        //remember you can't advance on something after you've deleted it
        for(Node* p = head; p != nullptr;){
            Node* temp = p;
            p = p->next;
            delete temp;
        }
    }

    //calling a function which returns a list
    LinkedList<int> f() {
        LinkedList<int> a;
        a.addEnd(5);
        a.addEnd(4);
        return a;
    }

    //move constructor, instead of giving memory to a variable that is going to die anyways, just move its memory
    //can only use when you call a function and return by value
    LinkedList(LinkedList&& orig) {
        head = orig.head;
        return head;
    }

};

int main() {
    LinkedList<int> a;
    const int n = 10;
    for(int i = 0; i < n; i++) {
        a.addEnd(i);
        a.addStart(i);
    }
    a.removeStart();

    LinkedList<int> b = a;
    LinkedList<int> c = f();
}